vllm/v1/guided_decoding/__init__.py

```diff
@@ -0,0 +1,208 @@
+            self.vllm_config.scheduler_config.max_num_seqs, self.vocab_size)
+
+    def __getitem__(self, key: GuidedDecodingKey) -> Optional[Grammar]:
+        if key in self.request_key_to_grammar:
+            # Move accessed item to the end (most recently used)
+            value = self.request_key_to_grammar.pop(key)
+            self.request_key_to_grammar[key] = value
+            return value
+        return None
+
+    def remove_requests(self, request_ids: List[str]) -> None:
+        with self._requests_lock:
+            self.requests = {
+                req
+                for req in self.requests if req.request_id not in request_ids
+            }
+
+    def populate_cache(self, request: Request):
+        if not request.use_guided_decoding:
+            return False
+        grammar = self.request_key_to_grammar.get(request.guided_decoding_key)
+        if grammar:
+            request.grammar = copy.copy(grammar)
+            return False
+        request.grammar = self.cache(request)
+        return True
+
+    def cache(self, request: Request):
+        return self.executor.submit(self._executor_loop, request)
+
+    def _executor_loop(self, request: Request) -> Grammar:
+        key = request.guided_decoding_key
+        with self._requests_lock:
+            self.requests.add(request)
+        if key in self.request_key_to_grammar:
+            grammar = self.request_key_to_grammar[key]
+            return copy.copy(grammar)
+        grammar = self.initialize_grammar(key)
+        # If cache is full, remove the least recently used item
+        if len(self.request_key_to_grammar) >= self.max_cache_size:
+            self.request_key_to_grammar.popitem(last=False)
+        self.request_key_to_grammar[key] = grammar
+        return copy.copy(grammar)
+
+    def initialize_grammar(self, key: GuidedDecodingKey) -> Grammar:
+        # Note that the request was validated in the engine core client,
+        # so at this point we know it is a supported type of request.
+        #
+        # TODO: we still need to handle xgrammar compilation failures
+        request_type, grammar_spec = key
+
+        if request_type == GuidedDecodingOptions.json:
+            # TODO -- allow any_whitespace to be configurable
+            # pending merge of https://github.com/vllm-project/vllm/pull/12744
+            ctx = self.compiler.compile_json_schema(grammar_spec,
+                                                    any_whitespace=False)
+        elif request_type == GuidedDecodingOptions.json_object:
+            ctx = self.compiler.compile_builtin_json_grammar()
+        elif request_type == GuidedDecodingOptions.grammar:
+            ctx = self.compiler.compile_grammar(grammar_spec)
+        else:
+            logger.error("Validation should have already occurred. "
+                         "Please file an issue.")
+            raise ValueError(
+                f"grammar is not of valid supported types. ({request_type!s})")
```

Comment:
Is it safe to raise an error here? Won't this crash the entire server?

Reply:
I think we need to have a better way to propagate exceptions up to the API server.

right now, iirc it would crash the engine.

Reply:
This should never happen. It's already validated in the input processor. See the log message just above it.

Reply:
It looks like there's an incoming error type (EngineGenerateError) that should handle this case of propagating non-fatal errors:
https://github.com/vllm-project/vllm/pull/14048

Reply:
Yeah. I started to try to build on top of that, but decided it'd be better to not block on that for now.